tensorboard/util/timing.py

```diff
@@ -0,0 +1,133 @@
+    ...     pass
+    ...
+    >>> def function_3():
+    ...     with log_latency("region_within_function"):
+    ...         pass
+    ...
+
+    Args:
+        region_name: An optional string name to associate with this
+            latency region. Optional if used as a decorator; required if
+            used as a context manager.
+        log_level: Optional integer logging level constant. Defaults to
+            `logging.INFO`.
+
+    Returns:
+        A decorated version of the input callable, or a dual
+        decorator/context manager with the input region name.
+    """
+
+    if log_level is None:
+        log_level = logging.INFO
+
+    if isinstance(region_name, str):
+        return _log_latency(region_name, log_level)
+    elif region_name is None:
+        return _LogLatencyDecorator(log_level)
+    else:
+        msg = "region_name must be a string or `None`; got: %r" % (region_name,)
+        if callable(region_name):
+            msg += " (hint: write `@log_latency()` instead of `@log_latency`)"
+        raise TypeError(msg)
+
+
+class _ThreadLocalStore(threading.local):
+    def __init__(self):
+        self.nesting_level = 0
+
+
+_store = _ThreadLocalStore()
+
+
+@contextlib.contextmanager
+def _log_latency(name, log_level):
+    if not logger.isEnabledFor(log_level):
+        yield
+        return
+
+    start_level = _store.nesting_level
+    try:
+        started = time.time()
+        _store.nesting_level = start_level + 1
+        indent = (" " * 2) * start_level
+        thread = threading.current_thread()
+        prefix = "%s[%x]%s" % (thread.name, thread.ident, indent)
+        _log(log_level, "%s ENTER %s", prefix, name)
+        yield
+    finally:
+        _store.nesting_level = start_level
+        elapsed = time.time() - started
+        _log(
+            log_level, "%s EXIT %s - %0.6fs elapsed", prefix, name, elapsed,
+        )
+
+
+def _log(log_level, msg, *args):
+    # Forwarding method to ensure that all logging statements
+    # originating in this module have the same line number; if the
+    # "ENTER" log is on a line with 2-digit number and the "EXIT" log is
+    # on a line with 3-digit number, the logs are misaligned and harder
+    # to read.
+    logger.log(log_level, msg, *args)
+
+
+class _LogLatencyDecorator:
+    """Decorator that raises a nice error if used as a context manager."""
+
+    def __init__(self, log_level):
+        self._log_level = log_level
+
+    def __enter__(self):
+        # Caller tried to use `with log_latency()` when they needed to
+        # write `with log_latency("my_region_name")`. Let them know.
+        raise TypeError(
+            "log_latency() cannot be used as a context manager unless "
+            "a region name is specified"
+        )
+
+    def __exit__(self, *exc_info):
+        pass
+
+    def __call__(self, func):
+        qualname = getattr(func, "__qualname__", func)
+        return _log_latency(qualname, self._log_level)(func)
```

Comment:
FWIW, this was a head-scratcher until I found this part of the contextlib docs:

> contextmanager() uses ContextDecorator so the context managers it creates can be used as decorators as well as in with statements. When used as a decorator, a new generator instance is implicitly created on each function call (this allows the otherwise “one-shot” context managers created by contextmanager() to meet the requirement that context managers support multiple invocations in order to be used as decorators).

This was non-obvious to me (especially the nuance about how it manages to recreate the context manager via internal machinery) and would be helpful to at least have a comment - or maybe just write out the decorator by hand to make it obvious that the context manager is being regenerated on each function call.  The fact that these context managers are also decorators feels mostly like a convenience when actually using them as decorators directly; we don't need that convenience (since we're using it in the context of our *own* decorator) so we could just be explicit about wrapping the decorated function in a `with` block.

Also, the fact that this is defined in an object that is *also* a context manager, although I see why it's there (to guard against incorrect usage of `with log_latency()`, feels extra confusing.  Reintroducing the `@log_latency` form of the decorator might make sense just so that we can make the first parameter to `log_latency()` mandatory and avoid this class of error entirely, even if it makes the contract slightly more subtle?

Or alternatively, we could perhaps at least combine this context manager with the overall `log_latency()` and always go through this context-manager-decorator rather than taking the shortcut of returning `_log_latency(...)` as a context-manager-decorator sometimes and other times returning this object.  E.g. the following:

```python
class log_latency(object):
    def __init__(self, region_name=None, log_level=None, reusable=True):
        self._context_manager = _log_latency(region_name, log_level)

    def __enter__(self):
        if self._region_name is None:
            raise TypeError(...)
        self._context_manager.__enter__()
        return self

    def __exit__(self, *exc):
        return self._context_manager.__exit__(*exc)

    def __call__(self, func):
        qualname = getattr(func, "__qualname__", func)
        return _log_latency(qualname, self._log_level)(func)
```

Reply:
> Reintroducing the `@log_latency` form of the decorator might make
> sense just so that we can make the first parameter to `log_latency()`
> mandatory and avoid this class of error entirely, even if it makes the
> contract slightly more subtle?

Okay, done. The analogous error that we no longer have a nice message
for is `with log_latency:`, which seems less likely (more obviously
wrong at call site) and would even more hackery to fix (something like
`log_latency = _LogLatency()` or a metaclass). This is fine enough.

Reply:
Oh, meant to note—

> The fact that these context managers are also decorators feels mostly
> like a convenience when actually using them as decorators directly; we
> don't need that convenience (since we're using it in the context of
> our _own_ decorator) so we could just be explicit about wrapping the
> decorated function in a `with` block.

—in the revised version I’m still using this convenience, because it’s
simpler than doing the `functools.wraps`/`*args, **kwargs` dance
ourselves, but I’ve spelled it out a bit more to make it clearer. Agreed
that the `contextlib` magic is a bit magic.

Reply:
Thanks! This version is vastly easier for me to wrap my head around :)

(I guess `with log_latency:` will just raise an `AttributeError: __enter__` - but if Python 3.7 feels that's a good enough way to indicate "you tried to use something as a context manager and it wasn't one" then I think this is good enough for us.)